rotkehlchen/chain/evm/decoding/beefy_finance/decoder.py

```diff
@@ -0,0 +1,223 @@
+
+from .constants import (
+    CPT_BEEFY_FINANCE,
+    SUPPORTED_BEEFY_CHAINS,
+    TOKEN_RETURNED_TOPIC,
+)
+from .utils import query_beefy_vaults
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import Asset, CryptoAsset, EvmToken
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.types import FVal
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class BeefyFinanceCommonDecoder(DecoderInterface, ReloadableDecoderMixin):
+    """Decodes Beefy Finance vault transactions into structured history events.
+
+    Handles both direct vault interactions and zap contract operations for
+    depositing and withdrawing from yield farming vaults.
+    """
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.vaults: set[ChecksumEvmAddress] = set()
+        self.zap_contract_address = SUPPORTED_BEEFY_CHAINS[self.evm_inquirer.chain_id]
+
+    def _process_beefy_events(
+            self,
+            events: list['EvmEvent'],
+            transaction: 'EvmTransaction',
+            from_address: ChecksumEvmAddress | None = None,
+            expected_amount: 'FVal | None' = None,
+            expected_asset: 'Asset | None' = None,
+    ) -> list['EvmEvent']:
+        """Core logic for processing Beefy finance events.
+
+        If `from_address` is provided, only processes events from that specific contract.
+
+        If `expected_amount` or `expected_asset` are provided, only matches receive events
+        with those exact values when multiple receives occur in transactions.
+        """
+        spend_events, receive_events = [], []
+        is_withdrawal = False
+
+        # process spend events and determine transaction type
+        for event in events:
+            if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE and
+                    (from_address is None or event.address == from_address)
+            ):
+                spend_asset: EvmToken | CryptoAsset
+                try:  # Determine if withdrawal by checking if vault tokens are being spent
+                    spend_asset = event.asset.resolve_to_evm_token()
+                    is_withdrawal = spend_asset.evm_address in self.vaults
+                except WrongAssetType:
+                    spend_asset = event.asset.resolve_to_crypto_asset()
+                    is_withdrawal = False
```

Comment:
you didn't adjust this as the note. Resolve to crypto asset once and check if it is evm and if it is evm then access the address. Resolving to crypto_asset will return an evm token if it is evm token unless I'm mistaken